### PR TITLE
manifest: synch MCUBoot to upstream of 7 10 2020

### LINKED ISSUE #28898: lwm2m_client can't start if mcuboot is enabled
**Describe the bug**
lwm2m_client can't start if build with CONFIG_BOOTLOADER_MCUBOOT=y

**To Reproduce**
`cd zephyrproject`
`
west build -b nrf52840dk_nrf52840 -s bootloader/mcuboot/boot/zephyr -d build-mcuboot
`
`
west flash -d build-mcuboot
`
`west build -b nrf52840dk_nrf52840 -s zephyr/samples/net/lwm2m_client -- -DCONFIG_BOOTLOADER_MCUBOOT=y 
-DCONFIG_MCUBOOT_SIGNATURE_KEY_FILE=\"bootloader/mcuboot/root-rsa-2048.pem\"
`
`
west flash
`

`
*** Booting Zephyr OS build zephyr-v2.4.0-64-g20df3ffc3dcd  ***
`
`
[00:00:00.269,836] <inf> mcuboot: Starting bootloader
`
`
[00:00:00.276,306] <inf> mcuboot: Primary image: magic=unset, swap_type=0x1, copy_done=0x3, image_ok=0x3
`
`
[00:00:00.286,590] <inf> mcuboot: Boot source: none
`
`
[00:00:00.292,236] <inf> mcuboot: Swap type: none
`
`
[00:00:00.377,624] <inf> mcuboot: Bootloader chainload address offset: 0xc000
`
`
[00:00:00.385,528] <inf> mcuboot: Jumping to the first image slot
`

**Expected behavior**
After control is passed to first image slot, APP main should be started:
`
*** Booting Zephyr OS build zephyr-v2.4.0-64-g20df3ffc3dcd  ***
`
`
[00:00:00.269,836] <inf> mcuboot: Starting bootloader
`
`
[00:00:00.276,306] <inf> mcuboot: Primary image: magic=unset, swap_type=0x1, copy_done=0x3, image_ok=0x3
`
`
[00:00:00.286,590] <inf> mcuboot: Boot source: none
`
`
[00:00:00.292,236] <inf> mcuboot: Swap type: none
`
`
[00:00:00.377,624] <inf> mcuboot: Bootloader chainload address offset: 0xc000
`
`
[00:00:00.385,528] <inf> mcuboot: Jumping to the first image slot
`
`
*** Booting Zephyr OS build zephyr-v2.4.0-64-g20df3ffc3dcd  ***
`
`
00:00:00.008,453] <inf> net_lwm2m_client_app: Run LWM2M client
`

**Impact**
showstopper

**Environment (please complete the following information):**
 - OS: Linux
 - Toolchain: gcc-arm-none-eabi-7-2018-q2-update
 - Zephyr version: 2.4.99
 - Commit SHA: 20df3ffc3dcd1de5a9ad45d713fc656ea88cad99

**Additional context**
If set CONFIG_LWM2M=n and comment all the code into lwm2m_client.c and left only 
`
LOG_INF(APP_BANNER);
`
into main(), then APP banner is shown.
Increasing CONFIG_MAIN_STACK_SIZE to 2048 or 4096, doesn't help.

If debug session is started with `west debug` and start it with 'continue', it starts ok:
`
*** Booting Zephyr OS build zephyr-v2.4.0-64-g20df3ffc3dcd  ***
`
`
uart:~$
`

### LINKED ISSUE #28921: MCUboot / smp_svr sample broken in 2.4.0
**Describe the bug**
since 2.4.0 the smp_svr sample is broken.
I have exactly followed the instructions, but it does not work, its also confirmed by sombody else.
Mcuboot runs fine, but as soon as it jumps to the smp_svr sample its just dead and nothing happens anymore. its also not advertising. This only happens when using the overlay-bt, overlay-bt-tiny works fine. It seems like it has something todo with the image size.
overlay-bt works fine with 2.3.0.

**To Reproduce**
just use west init and then west update to get a clean copy, then:

```
west build -b nrf52840dk_nrf52840 bootloader/mcuboot/boot/zephyr
west flash
rm -rf build
west build -b nrf52840dk_nrf52840 zephyr/samples/subsys/mgmt/mcumgr/smp_svr -- -DOVERLAY_CONFIG=overlay-bt.conf
west sign -t imgtool -- --key bootloader/mcuboot/root-rsa-2048.pem
west flash --hex-file build/zephyr/zephyr.signed.hex
```

**Expected behavior**
MCUboot loads the sample and runs

**Impact**
have to go back to 2.3.0

**Logs and console output**
```
*** Booting Zephyr OS build v2.4.0-rc3  ***
[00:00:00.376,495] [1B][0m<inf> mcuboot: Starting bootloader[1B][0m
[00:00:00.383,056] [1B][0m<inf> mcuboot: Primary image: magic=unset, swap_type=0x1, copy_done=0x3, image_ok=0x3[1B][0m
[00:00:00.393,493] [1B][0m<inf> mcuboot: Boot source: none[1B][0m
[00:00:00.399,261] [1B][0m<inf> mcuboot: Swap type: none[1B][0m
[00:00:00.678,283] [1B][0m<inf> mcuboot: Bootloader chainload address offset: 0xc000[1B][0m
[00:00:00.686,309] [1B][0m<inf> mcuboot: Jumping to the first image `slot[1B][0m`
```

**Environment (please complete the following information):**
 - OS: Ubuntu
 - Toolchain Zephyr SDK

**Additional context**
It seems like the problem has something todo with the image size, as I tried one of my own apps, and it also doesnt work (its way bigger than the sample)

### LINKED ISSUE #28933: mcuboot: Brick when using BOOT_SWAP_USING_MOVE and reset happens during images swap
**Describe the bug**
Interrupted swap of images when **swap_move** (`CONFIG_BOOT_SWAP_USING_MOVE=y`) strategy (default for nRF chips) is selected, can end up in "bricked" state. Possible workaround is to use **swap_scratch** (`CONFIG_BOOT_SWAP_USING_SCRATCH=y`) strategy, which doesn't seem to have this specific issue.

**To Reproduce**
1. `west build -d build_boot -b nrf52840dk_nrf52840 ../bootloader/mcuboot/boot/zephyr/`
2. `west flash -d build_boot`
3. `west build -b nrf52840dk_nrf52840 samples/subsys/mgmt/mcumgr/smp_svr/ -- -DOVERLAY_CONFIG=overlay-bt.conf`
4. `west sign -t imgtool --no-bin -H app.hex -- --key ../bootloader/mcuboot/root-rsa-2048.pem --version 1.0.0 --pad --confirm`
5. `west flash --hex-file app.hex`
6. `west sign -t imgtool --no-hex -B fake.bin -- --key ../bootloader/mcuboot/root-rsa-2048.pem --version 1.2.3`
7. `mcumgr --conntype ble --connstring peer_name=Zephyr image upload fake.bin`
8. `mcumgr --conntype ble --connstring peer_name=Zephyr image test 597f80daf86f4d5bbe0a23af28167ae1de07cbd441b1c2db862e8dff17cf2037`
```
Images:
 image=0 slot=0
    version: 1.0.0
    bootable: true
    flags: active confirmed
    hash: 096918ac22fbc49f7613ccabbd06853804c7422a7884239c212c7f70dd58976e
 image=0 slot=1
    version: 1.2.3
    bootable: true
    flags: pending
    hash: 597f80daf86f4d5bbe0a23af28167ae1de07cbd441b1c2db862e8dff17cf2037
Split status: N/A (0)
```
9. Press RESET button, wait ~2s, press RESET.
Log shows mcuboot messages after first (and second) reset:
```
$ cat /dev/ttyACM1 | ts '[%H:%M:%.S]'
[20:14:48.323953] *** Booting Zephyr OS build zephyr-v2.4.0-226-gb89548198d99  ***
[20:14:48.329925] [00:00:00.257,659] <inf> mcuboot: Starting bootloader
[20:14:48.339929] [00:00:00.264,129] <inf> mcuboot: Primary image: magic=good, swap_type=0x1, copy_done=0x3, image_ok=0x3
[20:14:48.346898] [00:00:00.274,414] <inf> mcuboot: Boot source: primary slot
[20:14:48.355929] [00:00:00.283,966] <inf> mcuboot: Swap type: test
[20:14:50.282960] *** Booting Zephyr OS build zephyr-v2.4.0-226-gb89548198d99  ***
[20:14:50.288973] [00:00:00.257,598] <inf> mcuboot: Starting bootloader
[20:14:50.298939] [00:00:00.264,068] <inf> mcuboot: Primary image: magic=good, swap_type=0x1, copy_done=0x3, image_ok=0x3
[20:14:50.305945] [00:00:00.274,322] <inf> mcuboot: Boot source: primary slot
```
10. Board never recovers.

**Expected behavior**
Swap of images should continue.

**Impact**
Remote firmware upgrade is not safe, i.e. can result in bricked devices in the field.

**Environment (please complete the following information):**
 - OS: Linux
 - Toolchain: Zephyr SDK 0.11.4
 - Commit SHA: dac6937cdacd4473f3e6d16964436cbf8c25b295 (2.4.99) + https://github.com/JuulLabs-OSS/mcuboot/pull/828 bugfix

**Additional context**
It seems that problems start here (https://github.com/JuulLabs-OSS/mcuboot/blob/aac99a14c9cdc8684fd4fbced130d3065a75ffea/boot/bootutil/src/loader.c#L1084):
```
        /*
         * If a swap was under way, the swap_size should already be present
         * in the trailer...
         */
        rc = boot_read_swap_size(image_index, &bs->swap_size);
        assert(rc == 0);
```
mcuboot reads `0xFFFFFFFF` after swap was interrupted, because this is what was contained originally in `app.hex` (signed and confirmed) image trailer and that was never updated by mcuboot. This value is passed as `copy_size` to `swap_run()` and results in infinite loop there (https://github.com/JuulLabs-OSS/mcuboot/blob/aac99a14c9cdc8684fd4fbced130d3065a75ffea/boot/bootutil/src/swap_move.c#L448):
```
    while (1) {
        sz += sector_sz;
        /* Skip to next sector because all sectors will be moved up. */
        g_last_idx++;
        if (sz >= copy_size) {
            break;
        }
    }
```

### PR DESCRIPTION
**Pull request by nvlsianpu**
Introduces what is in https://github.com/zephyrproject-rtos/mcuboot/pull/36

MCUBoot was synchronized up to:
https://github.com/JuulLabs-OSS/mcuboot/commit/296949e

Improvements:
- hardening against hardware level fault injection and timing attacks
- Abstract crypto primitives to simplify porting.
- boot: Add ram-load upgrade mode
- renamed single-image mode to single-slot mode

- kconfig: provide logic for setting key file, simplify prj.conf

- imgtool: Print image digest during verify
- imgtool: Add possibility to set confirm flag for hex files as well
- imgtool: --confirm implies --pad

- Added single-slot Zephyr-RTOS test build

fixes:
- bootutil: fix boostrapping in swap-move
- Disable HW stack protection (temporary hack)
- reset SPLIM registers before boot
- fixes build issue that occurs if CONF_FILE contains
multiple file paths instead of single file path.
- imgtool: Fix 'custom_tlvs' argument handling
- Turn off cache for Cortex M7 before chain-loading.
- hardening against hardware level fault injection and timing attacks

Fixes: #28921
Fixes: #28933
Fixes: #28898

### CODE REVIEW

**Review by mniestroj — COMMENTED** [review]
Update commit message to reference `296949e` mcuboot upstream. And add new fixes/improvements in changelog. Otherwise looks okay.

**Review by utzig — APPROVED** [review]

**Comment by utzig** [comment]
@nvlsianpu The commit message still needs to refer to the right SHA after the last push.

**Comment by nvlsianpu** [comment]
Added changelog.  @utzig can you check it?
cc @carlescufi

**Comment by utzig** [comment]
@nvlsianpu The first commit message still has the wrong SHA for latest upstream commit in mcuboot!

File: doc/releases/release-notes-2.5.rst [hunk 1]
```diff
@@ -225,6 +225,35 @@ HALs
 * HALs are now moved out of the main tree as external modules and reside in
   their own standalone repositories.
 
+MCUBoot
+*******
+
+* bootloader
+
+  * Added hardening against hardware level fault injection and timing attacks,
+    see :option:`CONFIG_BOOT_FIH_PROFILE_HIGH and similar kconfig options.
```
Comment (utzig) [inline doc/releases/release-notes-2.5.rst:234]: the `:option:` is missing an ending "`" after the Kconfig option's name.

**Comment by mniestroj** [comment]
Fixes: #28921
Fixes: #28933

**Comment by nvlsianpu** [comment]
@utzig @mniestroj 

Doc CI is failing just because an analyzer doesn't catch mcuboot Kconfig symbols (it is false positive then). 
Error: /home/runner/work/zephyr/zephyr/doc/_build/rst/doc/releases/release-notes-2.5.rst:233: WARNING: unknown option: CONFIG_BOOT_FIH_PROFILE_HIGH
/home/runner/work/zephyr/zephyr/doc/_build/rst/doc/releases/release-notes-2.5.rst:237: WARNING: unknown option: CONFIG_SINGLE_APPLICATION_SLOT
Error: Process completed with exit code 1.

Otherwise this PR is fine now.

**Review by utzig — APPROVED** [review]

**Comment by nvlsianpu** [comment]
@carlescufi ^^

**Review by mniestroj — APPROVED** [review]

**Comment by mniestroj** [comment]
@nvlsianpu You can modify PR description and add "Fixes: " to all issues that will be automatically closed by this PR.

**Review by carlescufi — APPROVED** [review]

File: doc/releases/release-notes-2.5.rst [hunk 2]
```diff
@@ -225,6 +225,35 @@ HALs
 * HALs are now moved out of the main tree as external modules and reside in
   their own standalone repositories.
 
+MCUBoot
+*******
+
+* bootloader
+
+  * Added hardening against hardware level fault injection and timing attacks,
+    see :option:`CONFIG_BOOT_FIH_PROFILE_HIGH` and similar kconfig options.
```
Comment (carlescufi) [inline doc/releases/release-notes-2.5.rst:234]: Don't use `:option:` here since this is not parsed

File: doc/releases/release-notes-2.5.rst [hunk 3]
```diff
@@ -225,6 +225,35 @@ HALs
...
+  * Introduced Abstract crypto primitives to simplify porting.
+  * Added ram-load upgrade mode (not enabled for zephy-rtos yet).
+  * Renamed single-image mode to single-slot mode,
+    see :option:`CONFIG_SINGLE_APPLICATION_SLOT`.
```
Comment (carlescufi) [inline doc/releases/release-notes-2.5.rst:238]: Don't use `:option:` here since this is not parsed

**Review by carlescufi — CHANGES_REQUESTED** [review]

**Comment by nvlsianpu** [comment]
`:option:` was removed

**Review by carlescufi — APPROVED** [review]

**Pull request merged**